README.md
# **Socket Audio Player**

- Written by: **Travis M. Moore**
- Latest version: **Version 1.1.0**
- Originally created: **February 28, 2023**
- Last edited: **March 27, 2023**
<br>
<br>

---

## Description
Audio player with server that accepts commands from "Vulcan Over Sockets" app. Necessary because Vulcan does not support a version of Python that works with ASIO. I know. 😒
<br>
<br>

---

## Getting Started

### Dependencies

- Windows 10 or greater (not compatible with Mac OS)

### Installing

- This is a compiled app; the executable file is stored on Starfile at: \\starfile\Public\Temp\MooreT\Custom Software
- Simply copy the executable file and paste to a location on the local machine
- Double click to start the app

### First Use
- Double-click to start the application for the first time.
- Go to **Tools-->Audio** Settings to enter a valid audio device ID.
- Go to **Server-->Start Server** to start the server.
<br>
<br>

---

## Audio Device
### Audio Device Selection
The Audio Settings window will display a list of available audio devices in 
tabular format (see image below). Navigate to **Tools-->Audio Settings** to open the Audio Settings window.

1. Use the "name" column to find the desired audio device (often 
the Hammerfall ASIO device in the clinical research labs). 
2. Ensure that the number of channels out in the "chans_out" column is not 0, 
and is sufficient for the desired number of audio channels/speakers. For example, to present audio from speaker 5, there must be at least 5 channels in the "chans_out" column.
3. After you have identified the appropriate audio device, enter its ID number 
from the "device_id" column into the "Audio Device ID" text entry box. 
4. Finally, click the SUBMIT button to save your selections. 

<img src="audio_settings_window.png" alt="Audio Settings Window image" width="600"/>

### Channel Routing
Channels are currently routed to speakers in order; the first channel of audio is routed to speaker 1, the second channel to speaker 2, etc. Currently the only way to control routing is to reorder the channels in the audio file. 
<br>
<br>

---

## Calibration
Navigate to **Tools-->Calibration** to open the calibration window (see image below).

<!-- <img src="./assets/images/calibration_window.png" alt="Calibration Window image" width="600"/> -->
<img src="calibration_window.png" alt="Calibration Window image" width="600"/>

### Calibration Stimulus Selection
There are two options when calibrating.

1. The Speech Task Controller contains its own white noise stimulus for general calibration. Select the "White Noise" button to use this stimulus.
2. The "Custom File" button allows for an existing calibration file to be loaded. For example, to present IEEE sentences, load in the IEEE calibration file. 

### Calibration Stimulus Playback
Clicking the PLAY button will present the calibration stimulus. You can set the level of the calibration stimulus using the "Level (dB)" text entry box. A level of -30 dB is relatively safe, and is the default value. Making the level more negative will decrease the presentation level. Making the level more positive will increase the presentation level. Be careful not to overdrive the speakers!
<br>
<br>

---

## Compiling from Source
```
pyinstaller --noconfirm --onefile --windowed --add-data "C:/Users/MooTra/Code/Python/vesta_sockets_audio/assets/cal_stim.wav;." --add-data "C:/Users/MooTra/Code/Python/vesta_sockets_audio/assets/README;README/"  "C:/Users/MooTra/Code/Python/vesta_sockets_audio/controller.py"
```
<br>
<br>

---

## Contact
Please use the contact information below to submit bug reports, feature requests and any other feedback. Thank you for using the Socket Audio Player!

- Travis M. Moore: travis_moore@starkey.com


CHANGELOG.md
# **Change Log**

---

## v1.1.0

Date: Mar 27, 2023

### Major Bug Fixes
1. Fixed bug where entering a level of 0 would be read as None. This resulted in normalizing instead of applying the level (if statement checking for level evaluated to None). The effect was a slow increase in level as the entered value approached 0, then a big jump at 0 (i.e., normalizing to full scale), then a return to a lower level above 0.
<br>
<br>

---

## v1.0.1

Date: Mar 17, 2023

### Minor Features
1. Cleaned up messages printed to console to be more legible.
<br>
<br>

---

## v1.0.0

Date: Mar 09, 2023

### Initial Release
1. An audio player with server for receiving instructions from the Vesta Over Sockets app. This is necessary because Vulcan requires Python 3.6 (32-bit), which does not work with the ASIO driver in the labs. 
<br>
<br>


menus/mainmenu.py
""" Main menu class for Vesta Audio
"""

# Import GUI packages
import tkinter as tk
from tkinter import messagebox

class MainMenu(tk.Menu):
    """ Main Menu
    """
    # Find parent window and tell it to 
    # generate a callback sequence
    def _event(self, sequence):
        def callback(*_):
            root = self.master.winfo_toplevel()
            root.event_generate(sequence)
        return callback
    

    def _bind_accelerators(self):
        self.bind_all('<Control-q>', self._event('<<FileQuit>>'))


    def __init__(self, parent, **kwargs):
        super().__init__(parent, **kwargs)

        font_size = 16


        #############
        # File menu #
        #############
        file_menu = tk.Menu(self, tearoff=False)
        # file_menu.add_command(
        #     label="Session...",
        #     command=self._event('<<FileSession>>')
        # )
        #file_menu.add_separator()
        file_menu.add_command(
            label="Quit",
            command=self._event('<<FileQuit>>'),
            accelerator='Crtl+Q'
        )
        self.add_cascade(label='File', menu=file_menu, font=("", font_size))


        ###############
        # Server Menu #
        ###############
        server_menu = tk.Menu(self, tearoff=False)
        server_menu.add_command(
            label="Start Server",
            command=self._event('<<ServerStartServer>>')
        )
        # Add Server menu to the menubar
        self.add_cascade(label="Server", menu=server_menu, font=("", font_size))


        ############## 
        # Tools menu #
        ##############
        tools_menu = tk.Menu(self, tearoff=False)
        tools_menu.add_command(
            label='Audio Settings...',
            command=self._event('<<ToolsAudioSettings>>')
        )
        tools_menu.add_separator()
        tools_menu.add_command(
            label='Calibration...',
            command=self._event('<<ToolsCalibration>>')
        )
        # Add Tools menu to the menubar
        self.add_cascade(label="Tools", menu=tools_menu, font=("", font_size))


        #############
        # Help menu #
        #############
        help_menu = tk.Menu(self, tearoff=False)
        help_menu.add_command(
            label='About...',
            command=self.show_about
        )
        help_menu.add_command(
            label='Help...',
            command=self._event('<<Help>>')
        )
        # Add help menu to the menubar
        self.add_cascade(label="Help", menu=help_menu, font=("", font_size))

        # Bind shortcut keys
        self._bind_accelerators()

    ##################
    # Menu Functions #
    ##################
    # HELP menu
    def show_about(self):
        """ Show the about dialog """
        about_message = 'Socket Audio Player'
        about_detail = (
            'Written by: Travis M. Moore\n'
            'Version 1.1.0\n'
            'Created: Feb 28, 2023\n'
            'Last edited: March 27, 2023'
        )
        messagebox.showinfo(
            title='About',
            message=about_message,
            detail=about_detail
        )
